GCP Multi-MachinePool support with single character leases.

dgoodwin opened the pull request:
Implement a leasing system to support multiple machinepools. "w" is reserved for the "worker" pool from the installer. "m" is reserved for masters to avoid colliding with master machines, even though they are not represented by a MachineSet yet.

GCP will be limited to 35 machine pools. (a-z0-9 minus "m")

dgoodwin commented:
Need to review code myself and do real world testing still so a bit rough.

dgoodwin commented:
Wondering if we should add labels or spec/status fields to help link MachineSets to their MachinePool.

staebler commented:
> Wondering if we should add labels or spec/status fields to help link MachineSets to their MachinePool.

I am not exactly following what you are proposing here. Does any of the following cover it.
1. We currently set the `hive.openshift.io/machine-pool` label on the MachineSets that Hive manages. The value of the label is the pool name.
2. The MachinePool status has the names of all of the MachineSets that comprise the MachinePool.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -1,13 +1,21 @@
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Put these k8s imports with the other k8s imports below.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -17,25 +25,32 @@ import (
 
 // NewGCPActuator is the constructor for building a GCPActuator
-func NewGCPActuator(gcpCreds *corev1.Secret, logger log.FieldLogger) (*GCPActuator, error) {
+func NewGCPActuator(client client.Client, gcpCreds *corev1.Secret, logger log.FieldLogger) (*GCPActuator, error) {
```
Did you intend to store `client` in the created actuator?

dgoodwin replied:
I did yes. Is there a reason not to? Viewing it as an extension of the actuator and it needs several client calls now, seemed more logical than passing it around several times.

staebler replied:
No. What I meant was that you are not setting the client when I think that you meant to.

dgoodwin replied:
LOL that's a good one, thanks. Would have hit that fast in real world  testing.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -84,7 +105,12 @@ func (a *GCPActuator) GenerateMachineSets(cd *hivev1.ClusterDeployment, pool *hi
+	// so we need to fake it to get it to use our leaseChar.
+	computePool.Name = string(leaseChar)
+	// Assuming all machine pools are workers at this time.
+	role := "worker"
```
```
const role = worker
```

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+	logger.Debugf("obtaining lease for pool: %s", pool.Name)
+
+	cd := &hivev1.ClusterDeployment{}
+	err := a.client.Get(context.TODO(), types.NamespacedName{Namespace: pool.Namespace, Name: pool.Spec.ClusterDeploymentRef.Name}, cd)
```
Why do we need to fetch the clusterdeployment? It was passed in to the `GenerateMachineSets` function which calls this function.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+	}
+
+	leases := &hivev1.MachinePoolNameLeaseList{}
+	// TODO: pass ^^ to find command?
```
What does this comment mean?

dgoodwin replied:
passing the lease list to the function that also does the same query later, which you caught with a comment there as well.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+
+	leases := &hivev1.MachinePoolNameLeaseList{}
+	// TODO: pass ^^ to find command?
+	err = a.client.List(context.TODO(), leases, client.UseListOptions(&client.ListOptions{Namespace: pool.Namespace}))
```
Replace
```
client.UseListOptions(...)
```
with
```
client.InNamespace(pool.Namespace)
```

staebler replied:
Add a label to the lease so that you can limit the list to only leases for the clusterdeployment via the `client.MatchingLabels` function.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+// for use in the name of the machine pool. We are severely restricted on name lengths on GCP
+// and effectively have one character of flexibility with the naming convention originating in
+// the installer.
+func (a *GCPActuator) obtainLease(pool *hivev1.MachinePool, logger log.FieldLogger) (rune, error) {
```
Return a `string` rather than a `rune`. Returning a `rune` (1) unnecessarily limits what this function can do in the future and (2) provides no benefit over a `string` as it will be converted to a `string` by the caller.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+	}
+	logger.Debugf("found %d leases in namespace", len(leases.Items))
+	for _, l := range leases.Items {
+		for _, ownerRef := range l.OwnerReferences {
```
Do not use owner references as authoritative: Use a label instead.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+	}
+
+	// Pool does not have a lease yet, lookup all currently available lease chars
+	availLeaseChars, err := a.findAvailableLeaseChars(cd)
```
Why not pass in the `MachinePoolNameLeaseList` to `findAvailableLeaseChars` rather than listing again?

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+
+	// If the pool.Spec.Name == "worker", we want to preseve this MachinePool's "w" character that
+	// the installer would have selected so we do not cycle all worker nodes. In theory, multiple
+	// MachinePools could specify this in their spec.Name, in which case we'll start erroring
```
This is not true. The pool name is unique for a given clusterdeployment because the name of the MachinePool must be $CLUSTERDEPLOYMENTNAME-$POOLNAME.

dgoodwin replied:
Ah yes! Thanks.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+	// MachinePools could specify this in their spec.Name, in which case we'll start erroring
+	// because the first one already grabbed "w".
+	var leaseRune rune
+	if pool.Spec.Name == "worker" {
```
I would rather special-case this earlier rather than leasing "w". In other words, if the pool name is "worker", then automatically use "w" for the lease char, do not look through existing leases, and do not create a lease. Remove "w" from the list of valid lease chars.

dgoodwin replied:
The idea here was to error if two pools had spec.name = worker, but if that's not possible then I'm with you.

staebler replied:
I don't know that we need extra protection for this specific to the GCP case. There is already protection in the admission web hook validators that rejects duplicates, and that is generic for all platforms.

staebler replied:
Ah, I misread your response. We are on the same page. Disregard my response.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+		// multiple machine pools at the same time. In this case the subsequent attempts to create
+		// the lease will fail due to name collisions and re-reconcile.
+		logger.Debug("selecting random lease char from available")
+		rand.Seed(time.Now().UnixNano())
```
I think that you want this already seeded rather than seeding anew on every use.

dgoodwin replied:
Whoops, moved to NewActuator.

staebler replied:
I would rather do it in `cmd/manager/main.go` or in an `init()` function.

dgoodwin replied:
Moved to main.go.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+
+	leaseName := fmt.Sprintf("%s-%s", cd.Spec.ClusterMetadata.InfraID, string(leaseRune))
+	// Attempt to claim the lease:
+	l := &hivev1.MachinePoolNameLease{
```
Don't we need to add the owner reference? And, preferably, the labels.

In addition to the owner reference to the MachinePool alluded to elsewhere in the code, there should also be an owner reference to the ClusterDeployment. We want the lease to expire when either the MachinePool or the ClusterDeployment are deleted.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+	availChars := stringToRuneSet(validLeaseChars)
+
+	leases := &hivev1.MachinePoolNameLeaseList{}
+	err := a.client.List(context.TODO(), leases, client.UseListOptions(&client.ListOptions{Namespace: cd.Namespace}))
```
`client.InNamespace`

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+	leases := &hivev1.MachinePoolNameLeaseList{}
+	err := a.client.List(context.TODO(), leases, client.UseListOptions(&client.ListOptions{Namespace: cd.Namespace}))
+	if err != nil {
+		return []rune{}, err
```
There is no need to allocate an array here or in any of the other returns where the function is returning an error.
```
return nil, err
```

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+			return []rune{}, fmt.Errorf("MachinePoolNameLease %s has unexpected owner ref: %v", lease.Name, lease.OwnerReferences[0])
+		}
+
+		// TODO: should we store the cluster deployment infra ID on the lease to make this unnecessary?
```
Yes. It should be stored in a label, although I don't follow why the infra ID is necessary and not just the clusterdeployment name.

But, we still need to check that the name of the lease is in the correct format. We are relying on that name to provide the uniqueness of the lease, so it **must** be correct.

dgoodwin replied:
Adding a check that everything but the last char matches infraid- which should cover us in all cases. (and a test)

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+		delete(availChars, rune(char))
+	}
+
+	keys := make([]rune, 0, len(availChars))
```
```
keys := make([]rune, len(availChars))
i := 0
for k := range availChars {
  keys[i] = k
  i++
}
```

dgoodwin commented:
Updated with everything but the expectations, I'll work off your past PRs for that first chance I get.

twiest commented:
@dgoodwin hey, above you said this PR was "rough". It also needs a rebase.

Are you looking for a PR review or do you want some more time to get it ready first?

dgoodwin commented:
Working on implementing the expectations but I'm hitting some things that I don't understand.

If I'm following the problem is reconciling a new pool, we list all leases in this namespace, find none owned by our pool, and create a new one. We then return from the obtainLease function but proceed to create the remote machine pools, assuming that our lease is now in place.

When combined with your comment that we *still* would need to handle two leases for one pool even with expectations I don't think we've implemented leases properly, and cleaning this up could be a pretty serious problem. 

First question would be how is this still possible with expectations?

Second question would be if there is some way to better approach this. If we stored the lease character onto status as soon as we obtain it, would this bypass the cache creation problem? Because frankly I'd massively prefer that to expectations plus handling whatever scenario could lead to two leases for one pool.

dgoodwin commented:
@twiest needs more time, I've done the rebase locally but working on the expectations problem first. Rest of the code is ok for review though.

dgoodwin commented:
Clarification from slack, expectations *should* prevent there ever being two leases for one pool. @staebler 's point was more in regards to defensive programming. I am working on the expectations and will give some thought to whether or not we can reasonably handle two leases for one pool. (or should just start erroring and let an admin sort it out)

dgoodwin commented:
/test e2e-gcp

dgoodwin commented:
I think this is ready for review, updated with expectations and tested to be working. I do need to go make sure we're hitting this as hard as other platforms in e2e.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +195,145 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+// for use in the name of the machine pool. We are severely restricted on name lengths on GCP
+// and effectively have one character of flexibility with the naming convention originating in
+// the installer.
+func (a *GCPActuator) obtainLease(pool *hivev1.MachinePool, cd *hivev1.ClusterDeployment, logger log.FieldLogger) (string, bool, error) {
```
Either (1) add details in the function description about what the second return value is or (2) name the return values.

dgoodwin replied:
Vars named.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +195,145 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+				// the last character.
+				expectedLeasePrefix := fmt.Sprintf("%s-", cd.Spec.ClusterMetadata.InfraID)
+				if l.Name[0:len(l.Name)-1] != expectedLeasePrefix {
+					return "", false, fmt.Errorf("lease %s did not match expected lease name format (%s[CHAR])", l.Name, expectedLeasePrefix)
```
Should Hive do something to resolve this?

dgoodwin replied:
I'm not sure what it would be. I don't think I'd dare delete it and let it create a new one as we might nuke a machine set. It felt like a pretty fringe/unexpected thing to happen, and a fair candidate to refuse to reconcile and require manual intervention from an admin who can figure out what happened.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -1,62 +1,111 @@
 
+const (
+	// Omit m, the installer used this for the master machines.
+	validLeaseChars = "abcdefghijklnopqrstuvwxyz0123456789"
```
The `w` character needs to be removed from the valid lease chars.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +195,145 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+		if err != nil {
+			return "", false, err
+		}
+		logger.Debugf("availLeaseChars: %v", availLeaseChars)
```
I do not think that this will print anything meaningful. I think it will only print the address of the slice--not the contents.

dgoodwin replied:
It printed the contents of the slice but not as characters. Removing as that's not useful either. :)

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +195,145 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+	a.expectations.ExpectCreations(types.NamespacedName{Namespace: pool.Namespace, Name: pool.Name}.String(), 1)
+	err = a.client.Create(context.TODO(), l)
+	if err != nil {
+		return "", false, err
```
If there is an error, you need to remove the expected creation.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +195,145 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+		},
+	}
+	logger.Debug("adding expectation for lease creation for this pool")
+	a.expectations.ExpectCreations(types.NamespacedName{Namespace: pool.Namespace, Name: pool.Name}.String(), 1)
```
It smells to me a bit that the responsibility for expectations is split between the controller and the actuator. I can live with it for now, but I reserve the right to revisit it at some point.

dgoodwin replied:
Yeah the thought process here was making all of this, leases and expectations, available to another actuator if they wanted to make use of it. I'm open to changing it in future though as I'm not sure this is the best either.

dgoodwin commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -1,62 +1,111 @@
+	}
+
+	if leaseCreated {
+		return nil, false, fmt.Errorf("new machine pool name lease created, waiting until creation is observed")
```
You're quite right, I did this both ways, put the proceed flag back in, and then forgot to take out the error.

dgoodwin commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +195,145 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+func lookupProjectID(kubeClient client.Client, cd *hivev1.ClusterDeployment, logger log.FieldLogger) (string, error) {
+
+	credsSecret := &corev1.Secret{}
+	if err := kubeClient.Get(context.Background(), types.NamespacedName{Namespace: cd.Namespace, Name: cd.Spec.Platform.GCP.CredentialsSecretRef.Name}, credsSecret); err != nil {
```
Awesome, missed that, thanks!

dgoodwin commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +195,145 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+		logger.Debugf("checking lease: %s", l.Name)
+		for _, ownerRef := range l.OwnerReferences {
+			logger.Debugf("checking owner ref: %v", ownerRef)
+			if ownerRef.APIVersion == "hive.openshift.io/v1" &&
```
Good idea, I will add the label to the selector and then skip the whole for loop.

dgoodwin commented:
Review feedback addressed and e2e testing enabled for GCP machinesets, which got a little bigger than I'd have liked. I was having a hard time seeing what was going on when waiting for timeouts so I reworked the tests to use logrus, added -v in the Makefile targets, and switched to using testify instead of just t.Fail.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -200,37 +201,35 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+		// Ensure the lease name is in the format we expect, we know everything up to
+		// the last character.
+		expectedLeasePrefix := fmt.Sprintf("%s-", cd.Spec.ClusterMetadata.InfraID)
+		if l.Name[0:len(l.Name)-1] != expectedLeasePrefix {
```
The `0` can be omitted.

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -200,37 +201,35 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+
+		// Ensure the lease name is in the format we expect, we know everything up to
+		// the last character.
+		expectedLeasePrefix := fmt.Sprintf("%s-", cd.Spec.ClusterMetadata.InfraID)
```
I think this can be made a little easier to read in the following way. Take it or leave it. Your choice.

```
leaseChar := l.Name[len(l.Name)-1:]
expectedLeaseName := fmt.Sprintf("%s-%s", cd.Spec.ClusterMetadata.InfraID, leaseChar)
if expectedLeastName != l.Name {
  return "", false, fmt.Errorf(...)
}
return leaseChar, false, nil
```

staebler commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -200,37 +201,35 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+			return "", false, fmt.Errorf("lease %s did not match expected lease name format (%s[CHAR])", l.Name, expectedLeasePrefix)
 		}
+
+		return string(l.Name[len(l.Name)-1]), false, nil
```
```
return l.Name[len(l.Name)-1:], false, nil
```

staebler commented on Makefile:
```diff
@@ -54,11 +54,11 @@ test-e2e:
 .PHONY: test-e2e-postdeploy
 test-e2e-postdeploy:
-	go test -timeout 0 -count=1 ./test/e2e/postdeploy/...
+	go test -v -timeout 0 -count=1 ./test/e2e/postdeploy/...
```
This will fill up the ci e2e logs with test output for successful tests, which is generally not what we want.

dgoodwin replied:
I don't think it's going to actually fill anything, just add some context but I'm fine with removing and going back to only on failures.

staebler commented on test/e2e/common/machinepoolnamelease.go:
```diff
@@ -0,0 +1,71 @@
+)
+
+// WaitForMachinePoolNameLease waits for a lease to appear in the given namespace for the given pool.
+func WaitForMachinePoolNameLease(cfg *rest.Config, namespace, poolName string, timeOut time.Duration) (string, error) {
```
This function strikes me as a lot more complicated than it needs to be. Can we accomplish the same thing more easily with `clientwatch.UntilWithSync`?

dgoodwin replied:
It's done, maybe a little cleaner.

staebler commented on test/e2e/postinstall/machinesets/infra_test.go:
```diff
@@ -22,12 +25,19 @@ import (
 )
 
+const (
+	workerPoolName       = "worker"
```
Either
`workerPoolName` and `infraPoolName`
or
`workerMachinePoolName` and `infraMachinePoolName`.

staebler commented on test/e2e/postinstall/machinesets/infra_test.go:
```diff
@@ -41,59 +51,62 @@ func TestScaleMachinePool(t *testing.T) {
-		t.Fatal("worker machine pool does not exist")
-	}
+	pool := common.GetMachinePool(cd, workerPoolName)
+	require.NotNil(t, pool, "worker machine pool does not exist: ", workerPoolName)
```
My preference would be to use `NotNilf` here instead.

dgoodwin commented:
Updates pushed.

I am fine with removing the -v from e2e makefile targets but I am very much in favor of the switch to logrus. t.Log is a common problem for go programmers and surfaces especially in long running tests, you see nothing until the test ends. That is the situation I was in fixing gcp and it's a pain. It's a popular request to change in go itself, but they can't because they have to first print pass or fail, and things depend on that output, so they cannot change it now. 

logrus gets us output on the fly, and respects -v, you will still see nothing for passing tests and everything once they fail without it. Meanwhile developers can use -v and have some understanding of what's going on in a test that can take 5-10 minutes. It's objectively better and I can't think of a downside. We're also already using loggers in the e2e common code, and mixing that with t.Log is weird.

IMO we should not be using t.Log anywhere, and opt for loggers instead.

dgoodwin commented:
/test e2e-azure

dgoodwin commented:
Leaving WIP until there's a go-ahead to squash and tests are clear.

staebler commented:
> logrus gets us output on the fly, and respects -v, you will still see nothing for passing tests and everything once they fail without it. Meanwhile developers can use -v and have some understanding of what's going on in a test that can take 5-10 minutes. It's objectively better and I can't think of a downside. We're also already using loggers in the e2e common code, and mixing that with t.Log is weird.

Ah, yes, you are correct. Logrus does respect the `-v` flag. I was confused with my IDE experience where the `-v` flag is added automatically. I retract my objection to using logrus.

staebler commented on test/e2e/common/machinepoolnamelease.go:
```diff
@@ -0,0 +1,78 @@
+	waitContext, cancel := context.WithTimeout(context.Background(), timeOut)
+	defer cancel()
+
+	gvk, err := apiutil.GVKForObject(&hivev1.MachinePoolNameLease{}, scheme)
```
You do not need all the scheme and codec cruft.

```
hiveClient, err := hiveclient.NewForConfig(config)
if err != nil {
  return "", errors.Wrap(err, "could not create the Hive client")
}

restClient := hiveClient.HiveV1().RESTClient()
```

dgoodwin replied:
Excellent, thanks.

staebler commented on test/e2e/common/machinepoolnamelease.go:
```diff
@@ -0,0 +1,78 @@
+		nil,
+		func(event watch.Event) (bool, error) {
+			switch event.Type {
+			case watch.Added, watch.Modified, watch.Deleted:
```
I'm not sure that you need to consider `watch.Deleted`.

staebler commented on test/e2e/common/machinepoolnamelease.go:
```diff
@@ -0,0 +1,78 @@
+		func(event watch.Event) (bool, error) {
+			switch event.Type {
+			case watch.Added, watch.Modified, watch.Deleted:
+				lease, ok = event.Object.(*hivev1.MachinePoolNameLease)
```
I don't think that you need to do this part here. I would move it to after the `UnitlWithSync` call. If there is an object received in the event that is not a MachinePoolNameLease, then that is reason enough to fail the test.

```
event, err := clientwatch.UntilWithSync(
  ...
  func(event watch.Event) (bool, error) {
    return event.Type == watch.Added || event.Type == watch.Modified
  },
)
if err != nil {
  return "", errors.Wrap(err, "timeout waiting for MachinePoolNameLease")
}
lease, ok := event.Object.(*hivev1.MachinePoolNameLease)
if !ok {
  return "", errors.Fmt("expected a MachinePoolNameLease object but got a %T object instead", event.Object)
}
logger.Infof("found MachinePoolName lease: %s", lease.Name)
return lease.Name, nil
```

dgoodwin replied:
Looks better.

dgoodwin commented:
Feedback addressed, squashed into one commit. Removing WIP.

staebler commented on test/e2e/common/machinepoolnamelease.go:
```diff
@@ -0,0 +1,65 @@
+		options.LabelSelector = fmt.Sprintf("%s=%s", constants.MachinePoolNameLabel, poolName)
+	}
+
+	var lease *hivev1.MachinePoolNameLease
```
Can you remove these two declarations?

dgoodwin replied:
done

dgoodwin commented:
/test e2e-gcp
/test e2e-azure

dgoodwin commented:
One of the errors:

```
time="2020-03-05T01:39:43Z" level=error msg="running the apply command failed" controller=hive error="error when creating \"/var/cache/kubectl/apply-342300959\": clusterroles.rbac.authorization.k8s.io \"hive-admin\" is forbidden: user \"system:serviceaccount:hive:hive-operator\" (groups=[\"system:serviceaccounts\" \"system:serviceaccounts:hive\" \"system:authenticated\"]) is attempting to grant RBAC permissions not currently held:\n{APIGroups:[\"hive.openshift.io\"], Resources:[\"machinepoolnameleases\"], Verbs:[\"get\" \"list\" \"watch\"]}" stderr= stdout=
time="2020-03-05T01:39:43Z" level=error msg="error applying asset" asset=config/rbac/hive_admin_role.yaml controller=hive error="error when creating \"/var/cache/kubectl/apply-342300959\": clusterroles.rbac.authorization.k8s.io \"hive-admin\" is forbidden: user \"system:serviceaccount:hive:hive-operator\" (groups=[\"system:serviceaccounts\" \"system:serviceaccounts:hive\" \"system:authenticated\"]) is attempting to grant RBAC permissions not currently held:\n{APIGroups:[\"hive.openshift.io\"], Resources:[\"machinepoolnameleases\"], Verbs:[\"get\" \"list\" \"watch\"]}"
time="2020-03-05T01:39:43Z" level=error msg="error deploying Hive" controller=hive error="error when creating \"/var/cache/kubectl/apply-342300959\": clusterroles.rbac.authorization.k8s.io \"hive-admin\" is forbidden: user \"system:serviceaccount:hive:hive-operator\" (groups=[\"system:serviceaccounts\" \"system:serviceaccounts:hive\" \"system:authenticated\"]) is attempting to grant RBAC permissions not currently held:\n{APIGroups:[\"hive.openshift.io\"], Resources:[\"machinepoolnameleases\"], Verbs:[\"get\" \"list\" \"watch\"]}"
```

I've updated PR to add the permission to the operator role.

A number of test runs didn't make it past bootstrap, but sometimes they did and then hit this, last round all made it past so I'm hoping that was a flake in the CI system somewhere.

staebler commented:
> I've updated PR to add the permission to the operator role.

Do you think that there is any reason for hive_reader or hive_frontend to have at least GET permissions for machinepoolleasename?

dgoodwin commented:
I can't think of a reason they need it at this time, but we can add it in future if so.

Unit test failure is legit, there's a flake, if I run it enough times I can see that fail locally as well. Working on this now.

dgoodwin commented on pkg/controller/remotemachineset/gcpactuator.go:
```diff
@@ -142,3 +196,129 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+	err := a.client.List(context.TODO(), leases, client.InNamespace(pool.Namespace),
+		client.MatchingLabels(map[string]string{
+			constants.ClusterDeploymentNameLabel: cd.Name,
+			constants.MachinePoolNameLabel:       pool.Name,
```
Just found a bug with this refactor here, in limiting only to this pool, we later use this list to calc all the leases in use and thus always find nothing, resulting in possibility of hitting a char already in use error.

dgoodwin commented:
Operator still missing the permission, not sure what's up yet.

dgoodwin commented:
/retest

dgoodwin commented:
/test e2e-azure 
/test e2e-gcp

AWS passed

dgoodwin commented:
e2e and e2e-gcp are passing, azure failed for gateway timeouts, and previously for lack of IPs which I reported to dev productivity.

Needs an /lgtm, quick as possible. :)
